app/services/articles/feed.rb

```diff
@@ -36,12 +39,60 @@ def default_home_feed_and_featured_story(user_signed_in: false)
           limited_column_select.order("published_at DESC").limit(rand(15..80))
         hot_stories = hot_stories.to_a + new_stories.to_a
       end
+      hot_stories = rank_articles(hot_stories) if @user
       [featured_story, hot_stories]
     end
 
     def default_home_feed(user_signed_in: false)
       _featured_story, stories = default_home_feed_and_featured_story(user_signed_in: user_signed_in)
       stories
     end
+
+    def rank_articles(articles)
+      ranked_articles = articles.map do |article|
+        article_points = 0
+        article_points += score_followed_user(article)
+        article_points += score_followed_organization(article)
+        article_points += score_followed_tags(article)
+        article_points += score_randomness
+        article_points += score_language(article)
+        article_points += score_experience_level(article)
+        RankedArticle.new(article, article_points)
+      end
+      ranked_articles.sort { |a, b| b.score <=> a.score }.map(&:article)
+    end
+
+    def score_followed_user(article)
+      @user&.cached_following_users_ids&.include?(article.user_id) ? 1 : 0
+    end
+
+    def score_followed_tags(article)
+      @user.decorate.cached_followed_tags.select do |tag|
```

Comment:
Another micro-optimization, may not be necessary but we'd only be traversing `cached_followed_tags` tags once. This also pulls creating the tag list array out of the loop, which I think we should do either way:

```ruby
article_tags = article.decorate.cached_tag_list_array
@user.decorate.cached_followed_tags.sum_by |tag|
  article_tags.include?(tag.name) ? tag.points : 0
end
```

Reply:
Great suggestion. Also: TIL about `sum` (which look at the source does a `.map` and `.sum`, but reads much better!)

Reply:
Great suggestion. Also: TIL about `sum` (which look at the source does a `.map` and `.sum`, but reads much better!)

Reply:
Great suggestion. Also: TIL about `sum` (which look at the source does a `.map` and `.sum`, but reads much better!)